upcloud/managed_database.go

```diff
@@ -0,0 +1,693 @@
+	UseSysId        int           `json:"usesysid"`
+	WaitEvent       string        `json:"wait_event"`
+	WaintEventType  string        `json:"waint_event_type"`
+	XactStart       time.Time     `json:"xact_start"`
+}
+
+// MarshalJSON implements json.Marshaler
+func (m *ManagedDatabaseConnection) MarshalJSON() ([]byte, error) {
+	type alias ManagedDatabaseConnection
+	out := struct {
+		alias
+		QueryDuration float64 `json:"query_duration"`
+	}{}
+	out.alias = alias(*m)
+	out.QueryDuration = m.QueryDuration.Seconds()
+	return json.Marshal(&out)
+}
+
+// UnmarshalJSON implements json.Unmarshaler
+func (m *ManagedDatabaseConnection) UnmarshalJSON(d []byte) error {
+	type alias ManagedDatabaseConnection
+	in := struct {
+		alias
+		QueryDuration float64 `json:"query_duration"`
+	}{alias: alias(*m)}
+	if err := json.Unmarshal(d, &in); err != nil {
+		return err
+	}
+	*m = ManagedDatabaseConnection(in.alias)
+	m.QueryDuration = time.Duration(in.QueryDuration * float64(time.Second))
+	return nil
+}
+
+// ManagedDatabaseMaintenance represents the set time of week when automatic maintenance operations are allowed
+type ManagedDatabaseMaintenance struct {
+	DayOfWeek string `json:"dow"`
+	Time      string `json:"time"`
+}
+
+// ManagedDatabaseMetrics represents managed database service metrics
+//
+// Metrics are represented in chart form containing a set of columns and two-dimensional slice of rows.
+// The inner slice index corresponds a column. If the service consists of multiple nodes, each node gets their
+// own column in the chart.
+//
+// The first column is always a timestamp which denotes the timestamp for the recorded metric
+type ManagedDatabaseMetrics struct {
+	CPUUsage       ManagedDatabaseMetricsChartFloat64 `json:"cpu_usage"`
+	DiskUsage      ManagedDatabaseMetricsChartFloat64 `json:"disk_usage"`
+	DiskIOReads    ManagedDatabaseMetricsChartInt     `json:"diskio_reads"`
+	DiskIOWrite    ManagedDatabaseMetricsChartInt     `json:"diskio_writes"`
+	LoadAverage    ManagedDatabaseMetricsChartFloat64 `json:"load_average"`
+	MemoryUsage    ManagedDatabaseMetricsChartFloat64 `json:"mem_usage"`
+	NetworkReceive ManagedDatabaseMetricsChartInt     `json:"net_receive"`
+	NetworkSend    ManagedDatabaseMetricsChartInt     `json:"net_send"`
+}
+
+// ManagedDatabaseMetricsChartFloat64 represents a metric chart with float64 row values
+type ManagedDatabaseMetricsChartFloat64 struct {
+	ManagedDatabaseMetricsChartHeader
+	// Rows contains a slice of values per row. The inner slice has the same indexing as the
+	// ManagedDatabaseMetricsChartHeader.Columns
+	Rows [][]float64
+}
+
+// MarshalJSON implements json.Marshaler for the canonical form of metrics chart
+func (m ManagedDatabaseMetricsChartFloat64) MarshalJSON() ([]byte, error) {
+	var chart canonicalManagedDatabaseMetricChartForMarshal
+	chart.Data.Columns = make([]ManagedDatabaseMetricsColumn, 0, len(m.Columns)+1)
+	chart.Data.Columns = []ManagedDatabaseMetricsColumn{{Label: "time", Type: "date"}}
```

Comment:
Doesn't this re-create the slice, eg. the previous line with make is a no-op? Eg. should this really be `chart.Data.Columns = append(chart.Data.Columns, { Label: "time", Type: "date" })`?

Reply:
whoops :)

Reply:
implemented